### PR TITLE
Task/pull gfs data/2584

### PR DESCRIPTION
**Pull request by andrea-williams**
# Test Links:
[Landing Page](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/)
[MoreCast 2.0](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/morecast-2)
[Percentile Calculator](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/percentile-calculator)
[MoreCast](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/morecast)
[C-Haines](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/c-haines)
[FireBat](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator)
[FireBat bookmark](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/fire-behaviour-calculator?s=266&f=c5&c=NaN&w=20,s=286&f=c7&c=NaN&w=16,s=1055&f=c7&c=NaN&w=NaN,s=305&f=c7&c=NaN&w=NaN,s=344&f=c5&c=NaN&w=NaN,s=346&f=c7&c=NaN&w=NaN,s=328&f=c7&c=NaN&w=NaN,s=1399&f=c7&c=NaN&w=NaN,s=334&f=c7&c=NaN&w=NaN,s=1082&f=c3&c=NaN&w=NaN,s=388&f=c7&c=NaN&w=NaN,s=309&f=c7&c=NaN&w=16,s=306&f=c7&c=NaN&w=NaN,s=1029&f=c7&c=NaN&w=NaN,s=298&f=c7&c=NaN&w=NaN,s=836&f=c7&c=NaN&w=NaN,s=9999&f=c7&c=NaN&w=NaN)
[Auto Spatial Advisory (ASA)](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/auto-spatial-advisory)
[HFI Calculator](https://wps-pr-2648.apps.silver.devops.gov.bc.ca/hfi-calculator)

### CODE REVIEW

File: api/app/weather_models/process_grib.py [hunk 1]
```diff
@@ -183,6 +242,47 @@ def store_bounding_values(self,
...
+            u_wind_raster_band = dataset.GetRasterBand(GFS_000_HOURS_RASTER_BANDS.get('u_comp_wind_10m'))
+            v_wind_raster_band = dataset.GetRasterBand(GFS_000_HOURS_RASTER_BANDS.get('v_comp_wind_10m'))
+        elif (grib_info.prediction_timestamp - grib_info.model_run_timestamp).total_seconds() > 0:
+            tmp_raster_band = dataset.GetRasterBand(GFS_003_HOURS_RASTER_BANDS.get('tmp_tgl_2'))
+            rh_raster_band = dataset.GetRasterBand(GFS_003_HOURS_RASTER_BANDS.get('rh_tgl_2'))
+            u_wind_raster_band = dataset.GetRasterBand(GFS_003_HOURS_RASTER_BANDS.get('u_comp_wind_10m'))
+            v_wind_raster_band = dataset.GetRasterBand(GFS_003_HOURS_RASTER_BANDS.get('v_comp_wind_10m'))
+            precip_raster_band = dataset.GetRasterBand(GFS_003_HOURS_RASTER_BANDS.get('apcp_sfc_0'))
+
+        for (points, values) in self.yield_data_for_stations(tmp_raster_band):
```
Comment (conbrad) [inline api/app/weather_models/process_grib.py:269]: Lots of repetition here with these for blocks, I think a parameterized method could be pulled out here to reduce the duplication.

File: api/app/jobs/model_data.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+## GRIB Files
+
+All weather model produce outputs in GRIB file format. [Refer here for a description of what a GRIB file is](https://weather.gc.ca/grib/what_is_GRIB_e.html).
+
+A GRIB file contains one or more raster bands, in which weather data is presented in a grid format. Each raster band corresponds to one weather variable. (For example, temperature data points will be stored in a different raster band to wind speed data points.) In order to coalesce the points on the data grid with geographic locations (such as lat/long coordinates), a geographic coordinate transformation must be applied. This coordinate transformation is performed in our backend, using metadata supplied by the source of the GRIB file.
```
Comment (conbrad) [inline api/app/jobs/model_data.md:33]: This is a great overview, thank you. Are the coordinates of the data grid just (metric, value)? e.g. (temp, 21.2)?

For the geographic transform, is that based on a raster band in the grid file whose coordinates are (lat, long)?

File: api/app/jobs/model_data.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+For example, a small portion of the grid for a temperature raster band of a GRIB file might look like the image below. In this example, each point on the grid is spaced 5km apart. You can see that a fire weather station is located within one cell of this grid, where the 4 corners of the cell have temperature values of 3.1&deg;, 2.4&deg;, 2.7&deg;, and 2.7&deg;C. In order to approximate the predicted temperature at the weather station's exact location, we perform linear interpolation of the 4 temperature values listed above, where the weight of each data point is inversely proportional to the distance from the weather station to that point on the grid. Consequently, the 2.4&deg;C from the top right corner of the cell will have the highest weight as it is closest to the weather station, and the 2.7&deg;C from the bottom left corner will have the lowest weight as it is the furthest away.
+
+![Location-based linear interpolation for a weather station](../../../docs/images/Grid_wx_station.png)
```
Comment (conbrad) [inline api/app/jobs/model_data.md:108]: 🔥 🔥 🔥

File: api/app/jobs/model_data.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+
+Additionally, GFS model data is only relevant to Morecast users for the 12:00 PST (13:00 PDT) timeframe - this is the time of day that a fire weather forecast is issued for. Noon PST corresponds to 20:00 UTC, but the GFS model has no output for this hour. We therefore fetch data for the 18:00 and 21:00 UTC hours, and subsequently perform linear interpolation to calculate approximate weather values for 20:00 UTC.
+
+The filename convention that NOAA uses when publishing GFS model data follows this pattern (as an example):
+
+> gfs_4_20230214_0600_003.grb2
+
+where '20230214' is the date that the model was run (in YYYYMMDD format), '0600' is the time at which the model was run (06:00 UTC), and '003' indicates that the weather prediction is for 003 hours after the model run time, which would be 09:00 UTC on Feb 14, 2023.
+
+Therefore, to predict weather for 12:00 PST, we must fetch the following model run cycles:
```
Comment (conbrad) [inline api/app/jobs/model_data.md:85]: I'm not sure I understand the why for requiring these model run cycles for 12:00PST when above it says:
> Noon PST corresponds to 20:00 UTC, but the GFS model has no output for this hour. We therefore fetch data for the 18:00 and 21:00 UTC hours, and subsequently perform linear interpolation to calculate approximate weather values for 20:00 UTC.

File: api/app/jobs/noaa.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
+""" A script that downloads weather models from NCEI NOAA HTTPS data server
```
Comment (conbrad) [inline api/app/jobs/noaa.py:1]: Is our naming scheme for python modules/functions/classes based on the source then the model?

e.g here our module is `api/app/jobs/noaa.py`, NOAA being the data source, and we're getting the `GFS` model from this source, so we have the `GFS` class defined here.

File: api/app/jobs/noaa.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+                logger.error(
+                    'unexpected exception processing %s model run %d',
+                    self.model_type, hour, exc_info=exception)
+
+
+def process_models():
+    """ downloading and processing models """
+
+    # grab the start time.
+    start_time = datetime.datetime.now()
```
Comment (conbrad) [inline api/app/jobs/noaa.py:195]: Does this need to be using a specific timezone or do we just use the default on the pods where this runs?

File: api/app/weather_models/process_grib.py [hunk 7]
```diff
@@ -148,6 +163,53 @@ def yield_data_for_stations(self, raster_band: gdal.Dataset):
...
+        """
+        calc = math.atan2(v, u) * 180 / math.pi
+        return calc if calc > 0 else calc + 360
+
+    def yield_uv_wind_data_for_stations(self, u_raster_band: gdal.Dataset, v_raster_band: gdal.Dataset, variable: str):
+        """ Given a list of stations and 2 gdal datasets (one for u-component of wind, one for v-component
+        of wind), yield relevant data 
+
+        What the heck is going on here?! See
+        http://colaweb.gmu.edu/dev/clim301/lectures/wind/wind-uv
```
Comment (conbrad) [inline api/app/weather_models/process_grib.py:184]: 👍

File: api/app/weather_models/process_grib.py [hunk 8]
```diff
@@ -148,6 +163,53 @@ def yield_data_for_stations(self, raster_band: gdal.Dataset):
...
+                v_points, v_values = get_surrounding_grid(v_raster_band, x_coordinate, y_coordinate)
+
+                if variable == 'wdir_tgl_10':
+                    wind_dir_values = []
+                    for u, v in zip(u_values, v_values):
+                        wind_dir_values.append(self.calculate_wind_dir_from_u_v(u, v))
+                    if u_points == v_points:
+                        yield (u_points, wind_dir_values)
+                elif variable == 'wind_tgl_10':
+                    wind_sp_values = []
```
Comment (conbrad) [inline api/app/weather_models/process_grib.py:205]: Does `sp` stand for `speed`?

Comment (andrea-williams) [reply api/app/weather_models/process_grib.py:205]: it does, but I'll rename that for clarity

File: api/app/jobs/model_data.md (hunk 4 above)
Comment (dgboss) [reply api/app/jobs/model_data.md:85]: I just spent some time trying to wrap my head around this as well. I think some clarification would be helpful.

I'm guessing GFS gives predictions for specific times like ' the predicted temperature at 18:00 UTC will be 5*C'. Rather, I think it gives predictions at intervals, eg. the temperature 3 hours from now will be 5*C, 6 hours from now 7*C, etc.'

File: api/app/jobs/model_data.md (hunk 2 above)
Comment (andrea-williams) [reply api/app/jobs/model_data.md:33]: I'm not sure what the data grid overall looks like. AFAIK it's not possible to read from the datagrid directly. Instead we use gdal to retrieve a specific raster band from the data grid, where the id of the raster band corresponds to the weather variable we want. That then returns a 2D array of values for the weather variable. So ultimately we do something like `datagrid.GetRasterBand(581)` where 581 is the ID of the `TMP_TGL_2m` weather variable, and that returns something like

```
[[21.2, 21.4, 21.6, ...],
[20.9. 21.5, 23.4, ...],
...
]
```

I'll add a little bit more to the doc to explain this.

File: api/app/jobs/model_data.md (hunk 4 above)
Comment (andrea-williams) [reply api/app/jobs/model_data.md:85]: Added an extra sentence to the doc to explain why we use all 4 model run cycles.

File: api/app/jobs/noaa.py (hunk 5 above)
Comment (andrea-williams) [reply api/app/jobs/noaa.py:1]: lol I'm not sure what naming scheme we use in our project overall. We just had a script called `env_canada.py` for fetching the EC models, and I just followed along with that. Open to other suggestions though!

File: api/app/jobs/model_data.md (hunk 4 above)
Comment (andrea-williams) [reply api/app/jobs/model_data.md:86]: @dgboss  yes you are correct - it's intervals, not specific timestamps. Will try to explain that better in the doc.

File: api/app/weather_models/process_grib.py [hunk 9]
```diff
@@ -183,6 +245,47 @@ def store_bounding_values(self,
...
+        raster_band = dataset.GetRasterBand(1)
+        # Iterate through stations:
+        for (points, values) in self.yield_data_for_stations(raster_band):
+            self.store_bounding_values(
+                points, values, prediction_run, grib_info, session)
+
+    def process_noaa_grib_file(self, session: Session, dataset, grib_info: ModelRunInfo,
+                               prediction_run: PredictionModelRunTimestamp):
+        precip_raster_band = None
+        if (grib_info.model_run_timestamp - grib_info.prediction_timestamp).total_seconds() == 0:
```
Comment (dgboss) [inline api/app/weather_models/process_grib.py:260]: Could you please add a comment clarifying what this if statement is checking?

**Comment by andrea-williams** [comment]
Closing this PR because deployment to dev keeps failing. I think something in Openshift has been corrupted - possibly RC. Will open new PR from same branch.

**Pull request closed without merging**